compatibility Test coverage

adityabharadwaj198 opened the pull request:
* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
* feature


* **What is the current behavior?** (You can also link to an open issue here)
* Currently there are limited compatibility tests 


* **What is the new behavior (if this is a feature change)?**
* more compatibility tests added as part of this PR 


* **Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)
* No


* **Have unit tests been run against this PR?** (Has there also been any additional testing?)
* Sample run - https://github.com/marqo-ai/marqo/actions/runs/12542408189/job/34972376976#step:7:3587
* 


* **Related Python client changes** (link commit/PR here)
* NA


* **Related documentation changes** (link commit/PR here)
* NA


* **Other information**:
* NA


* **Please check if the PR fulfills these requirements**
- [ ] The commit message follows our guidelines
- [ ] Tests for the changes have been added (for bug fixes/features)
- [ ] Docs have been added / updated (for bug fixes / features)

adityabharadwaj198 committed:
Test coverage

adityabharadwaj198 committed:
Most tests run

adityabharadwaj198 committed:
Added one assertion statement

adityabharadwaj198 committed:
Made tests work

adityabharadwaj198 committed:
Moved all docker related logic to a seperate docker class & started using docker-py

adityabharadwaj198 committed:
Removed docker_manager.py for now

adityabharadwaj198 committed:
Fixed exception handling and changed log level

adityabharadwaj198 committed:
Fixed exception handling and changed log level

adityabharadwaj198 committed:
Merge remote-tracking branch 'origin/aditya/comp-test-coverage' into aditya/comp-test-coverage

# Conflicts:
#	tests/compatibility_tests/compatibility_test_runner.py

adityabharadwaj198 committed:
Removed docker related logic which was using command line

adityabharadwaj198 commented on tests/compatibility_tests/docker_manager.py:
Made this change to address 2 TODOs from an older PR:
1. https://github.com/marqo-ai/marqo/pull/1024/commits/5a9794d06352d68bed4c827a1c918282125225ae#r1841689970
2. https://github.com/marqo-ai/marqo/pull/1024/commits/5a9794d06352d68bed4c827a1c918282125225ae#r1842990800

adityabharadwaj198 commented on tests/backwards_compatibility_tests/compatibility_test_runner.py:
```diff
@@ -1,743 +0,0 @@
-
-    # Stop and remove the container if it exists
-    try:
-        subprocess.run(["docker", "rm", "-f", container_name], check=True)
-    except Exception as e:
-        logger.warning(f"Container {container_name} not found, skipping removal.")
-
-    # Prepare the docker run command
-    cmd = [
-        "docker", "run", "-d",
-        "--name", container_name,
-        "-p", "8882:8882",
-        "-e", "MARQO_ENABLE_BATCH_APIS=TRUE",
-        "-e", "MARQO_MAX_CPU_MODEL_MEMORY=1.6"
-    ]
-
-    # Handle version-specific volume mounting
-
-    # Mounting volumes for Marqo >= 2.9
-    # Use the provided volume for state transfer
-    volume_name = create_volume_for_marqo_version(version, volume_name)
-    logger.info(f"from_version volume name = {volume_name}")
-    if version >= marqo_transfer_state_version:
-        # setting volume to be mounted at /opt/vespa/var because starting from 2.9, the state is stored in /opt/vespa/var
-        cmd.extend(["-v", f"{volume_name}:/opt/vespa/var"])
-    else:
-        # setting volume to be mounted at /opt/vespa because before 2.9, the state was stored in /opt/vespa
-        cmd.extend(["-v", f"{volume_name}:/opt/vespa"])  # volume name will be marqo_2_12_0_volume
-
-    # Append the image
-    cmd.append(image_name)
-    logger.info(f"Running command: {' '.join(cmd)}")
-
-    try:
-        # Run the docker command
-        subprocess.run(cmd, check=True)
-        containers_to_cleanup.add(container_name)
-
-        # Follow docker logs
-        log_cmd = ["docker", "logs", "-f", container_name]
```
Made this change to address 2 TODOs from an older PR:

https://github.com/marqo-ai/marqo/commit/5a9794d06352d68bed4c827a1c918282125225ae#r1841689970
https://github.com/marqo-ai/marqo/commit/5a9794d06352d68bed4c827a1c918282125225ae#r1842990800

adityabharadwaj198 committed:
removing cleanup_containers() & cleanup_volumes() method

farshidz commented on tests/compatibility_tests/base_test_case/base_compatibility_test.py:
```diff
@@ -61,8 +63,11 @@ def load_results_from_file(cls):
-        cls.logger.debug(f"Results file deleted: {filepath}")
+        if filepath.exists():
+            filepath.unlink()
+            cls.logger.debug(f"Results file deleted: {filepath}")
```
I wonder if debug log level is excessive here. Do you think we will ever run this in debug mode? Is there a risk if we logged these at info level (i.e. print them all the time)?

adityabharadwaj198 replied:
we can run it in debug mode if debugging locally. If these are logged at info level then it would make the logs too verbose and it'd be difficult to find valuable information

farshidz commented on tests/compatibility_tests/add_or_replace_documents/test_add_documents.py:
```diff
@@ -0,0 +1,100 @@
+        "indexName": structured_index_name,
+        "type": "structured",
+        "model": "sentence-transformers/all-MiniLM-L6-v2",
+        "normalizeEmbeddings": False,
```
Non-normalized indexes need to be covered, but are unusual. If covering one, we should cover normalized

adityabharadwaj198 replied:
fixed

farshidz commented on tests/compatibility_tests/add_or_replace_documents/test_add_documents.py:
```diff
@@ -0,0 +1,100 @@
+    }]
+    @classmethod
+    def tearDownClass(cls) -> None:
+        cls.indexes_to_delete = [index['indexName'] for index in cls.indexes_to_test_on]
```
You could also do this in setupclass, right?

adityabharadwaj198 replied:
Now doing it in setupClass()

farshidz commented on tests/compatibility_tests/add_or_replace_documents/test_add_documents.py:
```diff
@@ -0,0 +1,100 @@
+        "model": "sentence-transformers/all-MiniLM-L6-v2",
+        "normalizeEmbeddings": False,
+        "allFields": [
+            {"name": "Title", "type": "text"},
```
You get the best coverage if you have all data types. That does mean you will need extra tests as some types were added in later versions. This comment applies to both index types.

adityabharadwaj198 replied:
Added more data types which were added as part of 2.0, 2.9, 2.12

farshidz commented on tests/compatibility_tests/add_or_replace_documents/test_add_documents.py:
```diff
@@ -0,0 +1,100 @@
+        try:
+            self.logger.debug(f'Feeding documents to {self.indexes_to_test_on}')
+            for index in self.indexes_to_test_on:
+                if index.get("type") is not None and index.get('type') == 'structured':
```
The problem with the way this is written is if the test fails for one index type, it doesn't run for the other. You need either subtests (if you can make that work here) or separate tests.

adityabharadwaj198 replied:
Subtests alone won't work here, because they just can be a wrapper above assertionErrors. Any other types of exceptions / errors and the subtest will fail and make the whole test fail. So I'm now using a try catch block + a subtest. Try catch block will be useful for any generic errors where as assertion errors can be logged by subtests

farshidz replied:
I'm not sure if you should worry about bugs in the test. It's fine for the whole test to fail if the problem is not an assertion. Try/catch to handle things that shouldn't be happening risks masking bugs

adityabharadwaj198 replied:
The test will eventually fail, this just enables it to run through all the test cases that are being iterated over using a for loop. If an exception happens maybe due to a bug, it will just be collected, and reported at the end. The logic here is such that if there's an exception we just report it at the end and fail the test.

farshidz commented on tests/compatibility_tests/add_or_replace_documents/test_add_documents.py:
```diff
@@ -0,0 +1,100 @@
+                try:
+                    all_results[index_name][doc_id] = self.client.index(index_name).get_document(doc_id)
+                except Exception as e:
+                    self.logger.error(
```
Shouldn't we fail here?

adityabharadwaj198 replied:
No point in failing a prepare mode test. Even if we fail here - when we run test mode we will not have the context of which prepare mode tests failed, and test mode test will fail again. So better to just let prepare mode run, and when the corresponding test mode test method runs - it will automatically fail.

farshidz commented on tests/compatibility_tests/add_or_replace_documents/test_add_documents.py:
```diff
@@ -0,0 +1,100 @@
+                try:
+                    expected_doc = stored_results[index_name][doc_id]
+                except KeyError as e:
+                    self.logger.error(f"The key {doc_id} doesn't exist in the stored results. Skipping the test for this document.")
```
Why are we not failing the test?

adityabharadwaj198 replied:
The test will now be failed, after running all the subTests, and collecting the failures / errors. At the end we have a check if there are any failures / errors we fail the test by doing self.fail()

farshidz commented on tests/compatibility_tests/create_index/test_create_index.py:
```diff
@@ -0,0 +1,36 @@
+        all_results = {}
+        try:
+            self.logger.debug(f"Creating index {self.index_name}")
+            self.client.create_index(index_name = self.index_name)
```
You're not testing structured

farshidz replied:
Actually saw the other one. Can we name this unstructured?

adityabharadwaj198 replied:
I want to stick to API names to lower confusion

farshidz commented on tests/compatibility_tests/create_structured_index/test_create_structured_index.py:
```diff
@@ -0,0 +1,66 @@
+class TestCreateStructuredIndex(BaseCompatibilityTestCase):
+    indexes_settings_to_test_on = [
+        {
+            "type": "structured",
```
There are field types that 2.0 didn't support. You need to cover those

adityabharadwaj198 replied:
Have covered them in tests corresponding to 2.2, 2.9, 2.12

farshidz commented on tests/compatibility_tests/create_structured_index/test_create_structured_index.py:
```diff
@@ -0,0 +1,66 @@
+                self.logger.debug(f"Printing actual_settings {actual_settings}")
+                self.logger.debug(f"Printing expected_settings {expected_settings.get(index_name)}")
+            except Exception as e:
+                raise Exception(f"Exception when getting index settings for index {index_name}") from e
```
This is a bit odd. We're not expecting an error here but if one does happen the test fails. I don't think you need try/except at all

adityabharadwaj198 replied:
try except is required so that even if one test fails the other can run. The issue here was that I was raising another exception anyway. I have changed that and now I let the other tests run before reporting exceptions

farshidz reviewed (changes requested):
There are a few problematic patterns. I'll do another pass once these have been addressed:
* Unnecessary try/except (then raise) catching generic Exception class. These are superfluous
* Different test cases in for loops, meaning if one test case fails, the others don't run
* Evolution of an API post-2.0 ignored, specifically for structured index field types but possibly also others

adityabharadwaj198 committed:
Addressing comments

adityabharadwaj198 committed:
Adding some more tests acc to post 2.0.0 API evolution

adityabharadwaj198 committed:
Fixing full_test_run for Rollback. Do I need to fix it for bw-comp also?

adityabharadwaj198 committed:
More tests to increase coverage. Contains all of data types for creating structured indexes and honors API evolution

adityabharadwaj198 committed:
More tests + fix some issues with test

adityabharadwaj198 committed:
Updating changed folder name for compatibility_tests, and adding a readme.md file

adityabharadwaj198 committed:
Fixing some test failures

adityabharadwaj198 committed:
Fixing some more test failures

adityabharadwaj198 committed:
Fixing some more test failures #2

adityabharadwaj198 committed:
Fixing some more test failures #3

adityabharadwaj198 committed:
Fixing one thing I committed by mistake + fixing delete API

adityabharadwaj198 committed:
Making changes to how we prepare the volume for rollback

adityabharadwaj198 committed:
Adding setupClass for test_vector_normalisation.py

adityabharadwaj198 committed:
Removing some comments and changing how we prepare volume to the old way without using docker-py

adityabharadwaj198 committed:
Merge branch 'mainline' into aditya/comp-test-coverage

Merging mainline into aditya/comp-test-coverage as it's quiet behind mainline and also possible contains different commits

adityabharadwaj198 committed:
Merge branch 'mainline' into aditya/comp-test-coverage

adityabharadwaj198 committed:
Fixing how we compare results in test_recommend.py. Printing a better exception trace in test_vector_normalisation.py and removing unused code from compatibility_test_runner.py

adityabharadwaj198 committed:
Merge remote-tracking branch 'origin/aditya/comp-test-coverage' into aditya/comp-test-coverage

adityabharadwaj198 committed:
Making test_vector_normalisation.py use a 384 dimension model called sentence-transformers/all-MiniLM-L6-v2 to fix error of

marqo.errors.MarqoWebError: MarqoWebError: MarqoWebError Error message: {'message': '[{"loc": ["__root__"], "msg": "Custom vector must have dimension of 768, but given vector is of length 384", "type": "value_error"}]', 'code': 'invalid_argument', 'type': 'invalid_request', 'link': ''}
